adapters/repos/db/vector/cache/sharded_lock_cache.go

```diff
@@ -353,3 +367,264 @@ func NewNoopCache(vecForID common.VectorForID[float32], maxSize int,
+	s.shardedLocks.RLock(docID)
+	vec := s.cache[docID][vecID]
+	s.shardedLocks.RUnlock(docID)
+
+	if vec != nil {
+		return vec, nil
+	}
+
+	return s.handleMultipleCacheMiss(ctx, docID, vecID)
+}
+
+func (s *shardedMultipleLockCache) Delete(ctx context.Context, docID uint64) {
+	panic("not implemented")
+}
+
+func (s *shardedMultipleLockCache) handleMultipleCacheMiss(ctx context.Context, docID uint64, vecID uint64) ([]float32, error) {
+	if s.allocChecker != nil {
+		// we don't really know the exact size here, but we don't have to be
+		// accurate. If mem pressure is this high, we basically want to prevent any
+		// new permanent heap alloc. If we underestimate the size of a vector a
+		// bit, we might allow one more vector than we can really fit. But the one
+		// after would fail then. Given that vectors are typically somewhat small
+		// (in the single-digit KBs), it doesn't matter much, as long as we stop
+		// allowing vectors when we're out of memory.
+		estimatedSize := int64(1024)
+		if err := s.allocChecker.CheckAlloc(estimatedSize); err != nil {
+			s.logger.WithFields(logrus.Fields{
+				"action": "vector_cache_miss",
+				"event":  "vector_load_skipped_oom",
+				"doc_id": docID,
+				"vec_id": vecID,
+			}).WithError(err).
+				Warnf("cannot load vector into cache due to memory pressure")
+			return nil, err
+		}
+	}
+
+	vec, err := s.multipleVectorForID(ctx, docID, vecID)
+	if err != nil {
+		return nil, err
+	}
+
+	atomic.AddInt64(&s.count, 1)
+	s.shardedLocks.Lock(docID)
+	s.cache[docID][vecID] = vec
+	s.shardedLocks.Unlock(docID)
+
+	return vec, nil
+}
+
+func (s *shardedMultipleLockCache) LockAll() {
+	s.shardedLocks.LockAll()
+}
+
+func (s *shardedMultipleLockCache) UnlockAll() {
+	s.shardedLocks.UnlockAll()
+}
+
+func (s *shardedMultipleLockCache) Prefetch(id uint64) {
+	panic("not implemented")
+}
+
+func (s *shardedMultipleLockCache) PrefetchMultiple(docID uint64, vecID uint64) {
+	s.shardedLocks.RLock(docID)
+	defer s.shardedLocks.RUnlock(docID)
+
+	prefetchFunc(uintptr(unsafe.Pointer(&s.cache[docID][vecID])))
+}
+
+func (s *shardedMultipleLockCache) PreloadMultiple(docID uint64, vecID uint64, vec []float32) {
+	s.shardedLocks.Lock(docID)
+	defer s.shardedLocks.Unlock(docID)
+
+	atomic.AddInt64(&s.count, 1)
+	s.cache[docID][vecID] = vec
+}
+
+func (s *shardedMultipleLockCache) Preload(docID uint64, vec []float32) {
+	panic("not implemented")
+}
+
+func (s *shardedMultipleLockCache) PreloadNoLock(docID uint64, vec []float32) {
+	panic("not implemented")
+}
+
+func (s *shardedMultipleLockCache) SetSizeAndGrowNoLock(size uint64) {
+	atomic.StoreInt64(&s.count, int64(size))
+
+	if size < uint64(len(s.cache)) {
+		return
+	}
+	newSize := size + MinimumIndexGrowthDelta
+	newCache := make([][][]float32, newSize)
+	for i := range newCache {
+		newCache[i] = make([][]float32, RelativeInitialSize)
+	}
+	copy(newCache, s.cache)
```

Comment:
`copy` only does a shallow copy, i.e. only the top level slice is copied.
What happens here is that you are only copying the pointers to the sub slices created the very first time the cache was used, while overwriting the pointer to the empty slices you have created above.
These empty slices are then never used and gc'ed eventually.